---
title: "PCHi-C EDA"
author: "Coleman Breen"
date: "2023-03-14"
output: html_document
---


# Summary

- Apply a threshold of 10 for cell-type specific CHiCAGO scores  
- Also threshold effect size (0.05)
- Run simulations separately  

# How Should We (If At All) Threshold Interactions?



```{r setup}

suppressPackageStartupMessages({
  library(data.table)
  library(tidyverse)
  library(GenomicRanges)
  library(rtracklayer)
  library(liftOver)
  library(org.Hs.eg.db)
  library(tibble)
  library(tidygraph)
  library(EnsDb.Hsapiens.v86)
})

IFILE <- "../../DataRaw/2023-02-14-Summaries-v6/pvals.bed"
PCHIC.FILE <- "../../DataReference/PCHi-C/PCHiC_peak_matrix_cutoff5.txt"
REF.CHAIN.PATH <- "../../DataReference/hg38ToHg19.over.chain"

```

```{r}
expression.genes <- readxl::read_xlsx("../../DataReference/DEGenes/2020-Shigemizu-AD-RNAseq-DEGenes.xlsx",
                                      skip = 1) %>%
  janitor::clean_names() 

nrow(expression.genes)
```

```{r ensdb}
db <- EnsDb.Hsapiens.v86
db.genes <- genes(db)

# Get valid gene symbols --------------------------------------------------
valid.symbols <- db.genes[db.genes$gene_biotype == "protein_coding"]$symbol


# All cell types available
cell.types <- c("Mon", "Mac0", "Mac1", "Mac2",
                "Neu", "MK", "EP", "Ery", "FoeT", "nCD4", "tCD4",
                "aCD4", "naCD4", "nCD8", "tCD8", "nB", "tB")
```


```{r filter_autosomes}
# PCHi-C Data Read-in and Munging -----------------------------------------


# Now the PCHI-C data (don't lift; keep in hg19)
# 1. Drop the cell types we don't care about (see `sub.cell.types` for types of interest)
# 2. Compute a max chicago score for remaining cells and subset
promcap.1 <- fread(PCHIC.FILE) %>% 
  dplyr::filter(baitChr %in% 1:22, oeChr %in% 1:22)
```


```{r filter_protein_coding}
# Don't consider non protein-coding genes
valid_row <- function(ix){
  any(str_split(promcap.1$baitName[ix], ";")[[1]] %in% valid.symbols)
}

# ix <- do.call(c, lapply(1:nrow(promcap.1), valid_row))
# promcap.2 <- promcap.1[ix, ]
promcap.2 <- promcap.1
```

# Filter on Protein Coding Only?
It seems like many of the interesting findings are **not** in protein coding genes. Don't filter (for now).

```{r pchic}

promcap.3 <- promcap.2 %>%
  dplyr::mutate(bait.id = paste0("chr", baitChr,":", baitStart, "-", baitEnd),
                oe.id = paste0("chr", oeChr,":", oeStart, "-", oeEnd))


# Summary stats
promcap.3$min.chicago <- apply(FUN=min, promcap.3[, ..cell.types], MARGIN=1)
promcap.3$med.chicago <- apply(FUN=median, promcap.3[, ..cell.types], MARGIN=1)
promcap.3$mean.chicago <- apply(FUN=mean, promcap.3[, ..cell.types], MARGIN=1)
```


We want to marginally examine promoter interactions. For a given cell type (e.g. naCD4), how many significant interactions are there? How many purported enhancers have 1+ DMPs in them? Does this pass a significance test?

```{r pchic}

filter_by_cell_type <- function(promcap, ct, chicago.threshold){
  ix <- promcap[[ct]] > chicago.threshold
  
  out <- makeGRangesFromDataFrame(promcap[ix, ],
                             seqnames.field = "oeChr",
                             start.field = "oeStart",
                             end.field = "oeEnd",
                             keep.extra.columns = T)
  seqlevelsStyle(out) <- "UCSC"
  out
}

```


```{r pvals}
chain <- import.chain(REF.CHAIN.PATH)

# 1. Read only the necessary columns for pvals
# 2. Conver to GRanges
# 3. Lift from hg38 to hg19
pvals.lifted.gr <- fread(IFILE) %>%
  dplyr::select(-c(p.from.zz, p.from.DSS, lfdr.from.zz)) %>%
  makeGRangesFromDataFrame(starts.in.df.are.0based = T,
                           keep.extra.columns = T) %>%
  liftOver(chain) %>%
  unlist()
```


```{r join_}
find_and_curate_overlaps <- function(dmps, pchic){
  # 
  overlaps <- findOverlaps(dmps, pchic)

  dmp.ix <- queryHits(overlaps)
  pchich.ix <- subjectHits(overlaps)

  # This is the data we want from the dmps
  a <- as.data.frame(dmps[dmp.ix])

  # Data from PCHi-C with some munging
  b <- as.data.frame(pchic[pchich.ix]) %>%
    dplyr::transmute(oe.chr = seqnames,
                     oe.start = start,
                     oe.end = end,
                     oe.id,
                     bait.chr = paste0("chr", baitChr),
                     bait.start = baitStart,
                     bait.end = baitEnd,
                     bait.id,
                     linked.gene = baitName,
                     dist,
                     !!!rlang::syms(cell.types))

  result <- cbind(a, b) %>%
    dplyr::mutate(Chr = as.numeric(stringr::str_remove(seqnames, "chr"))) %>%
    arrange(Chr, start)

  return(result)
}
```


```{r pvals}
get_unique_genes_from_overlaps <- function(result, filter = F){
  # filter subsets by VALID.IDS
  symbols <- unlist(stringr::str_split(result$linked.gene, ";"))

  if (filter){
    sort(unique(symbols[symbols %in% valid.symbols]))
  } else {
    sort(unique(symbols))
  }
}
```


```{r tally_function}
# Functions ------------------------------------------------------------
tally_overlaps <- function(dmps, pchic){
    
  # Query, subject
  overlaps <- findOverlaps(dmps, pchic)
  
  # queryHits gives the indices w.r.t. DMPs
  # subjectHIts gives indices w.r.t. enhancers
  # print(apply(as.data.frame(overlaps), FUN=max, MARGIN=2))
    
  out <- c(length(unique(queryHits(overlaps))), 
           length(unique(subjectHits(overlaps))))
  
  names(out) <- c("N.uniq.dmps", "N.uniq.enhancers")
  out
}

filter_by_cell_type(promcap.3, "EP", 10)
```

```{r pvals}
# Non-parameteric test ----------------------------------------------------
promcap.filt.gr <- filter_by_cell_type(promcap.3, "Neu", 20)
length(promcap.filt.gr)

#Threshold for DMP
alpha <- 0.05
delta <- 0.05

# Null: the number of DMPs residing in enhancers is more than CpGs being drawn randomly
dmps.gr <- pvals.lifted.gr[pvals.lifted.gr$lfdr.from.ss < alpha & abs(pvals.lifted.gr$pi.diff) > delta]
length(dmps.gr)


# We'll sample N.dmps number of CpGs
N.dmps <- length(dmps.gr)
tally <- tally_overlaps(dmps.gr, promcap.filt.gr)
tally
```


```{r simulate}
run_simulation <- function(B, N.dmps, cpg.loci, enhancers) {
  # B: number of simulations
  # N.dmps: number of DMPs (significant loci)
  # loci: granges with 
  
  # G: number of CpGs
  G <- length(cpg.loci)
  
  # Initialize empty matrix to store the number of 
  # unique DMPs and the number of unique 
  test.draws <- matrix(NA, nrow=B, ncol=2)
  
  for (b in 1:B){
  
    if (b %% 1000 == 0){print(paste0("Iteration: ", b))}
  
    # Sample G CpGs from all of the 25+ million possible
    ix <- sample(1:G, size = N.dmps, replace = F)
    random.dmps <- cpg.loci[ix]
  
    # Overlap with PCHi-C
    test.draws[b, ] <- tally_overlaps(random.dmps, enhancers)
  }
  out <- data.frame(test.draws)
  colnames(out) <- c("N.dmps.sim", "N.enhancers.sim")
  out
}
```

What if we threshold 
```{r iterate_thru_cells}
B <- 5000 # number of simulations
chicago.threshold <- 20

# Set seed before running on all
set.seed(919)

routine_by_celltype <- function(ct){
  promcap.filt.gr <- filter_by_cell_type(promcap.3, ct, chicago.threshold)
  tmp <- run_simulation(B, N.dmps, pvals.lifted.gr, promcap.filt.gr)
  tmp$cell.type <- ct
  tmp$N.enhancer <- length(promcap.filt.gr)
  tmp
}
  
sim.results <- do.call(rbind, lapply(cell.types, FUN=routine_by_celltype))
```

```{r}

test_stat_by_celltype <- function(ct){
  promcap.filt.gr <- filter_by_cell_type(promcap.3, ct, chicago.threshold)
  result <- tally_overlaps(dmps.gr, promcap.filt.gr)
  
  result$cell.type <- ct
  result$N.enhancers <- length(promcap.filt.gr)
  
  as.data.frame(result)
}


test.stats <- do.call(rbind, lapply(cell.types, FUN = test_stat_by_celltype))
test.stats
```

We'll need to compute p-values. Highlight plots if they're significant 

```{r}

empirical_pval <- function(simulated, observed){
  
  # Two-sided test
  left.mass <- sum(simulated <= -abs(observed) )
  right.mass <- sum(simulated >= abs(observed))
  
  return( (left.mass + right.mass) / length(simulated))
}


empirical_pval_by_ct <- function(ct){
  ss <- dplyr::filter(sim.results, cell.type == ct)$N.dmps.sim
  tt <- dplyr::filter(test.stats, cell.type == ct)$N.uniq.dmps
  
  empirical_pval(ss, tt)
}


p.df <- data.frame(
  pval = unlist(lapply(X=cell.types, FUN=empirical_pval_by_ct)),
  cell.type = cell.types
  )

```



```{r plot_sim_results}

# Group a bit better for plotting
ct.levels <- c("Mac0", "Mac1", "Mac2",
               "Mon", "FoeT", "MK",
               "aCD4", "naCD4", "nCD4", "tCD4",
               "nCD8", "tCD8", "nB", "tB",
               "Neu", "EP", "Ery")

alpha.ct.overlap <- 0.05

plot.hists <- sim.results %>% 
  ggplot(aes(x = N.dmps.sim)) +
  geom_histogram(aes(y = after_stat(density)), bins = 48) +
  facet_wrap(.~factor(cell.type, levels = ct.levels), ncol=5) +
  geom_vline(data=test.stats, aes(xintercept = N.uniq.dmps), color = "blue") +
  geom_rect(data = subset(sim.results, cell.type %in% p.df$cell.type[p.df$pval < alpha.ct.overlap]), 
                          fill = NA, colour = "orange", xmin = -Inf,xmax = Inf,
            ymin = -Inf,ymax = Inf) +
    geom_rect(data = subset(sim.results, cell.type %in% p.df$cell.type[p.df$pval < alpha.ct.overlap / nrow(p.df)]), 
                          fill = NA, colour = "red", xmin = -Inf,xmax = Inf,
            ymin = -Inf,ymax = Inf) +
  xlab("Number of unique DMPs residing in enhancer") +
  ylab("") +
  theme_minimal() +
  theme(axis.ticks = element_blank(), 
        axis.text.y = element_blank(),
        plot.background = element_rect(fill = "white", color = "white")) 


cowplot::save_plot("../../Figs/PCHi-C/histogram-simulation-by-ct.png", plot = plot.hists, base_width = 6, base_height = 6)
plot.hists


```


# Relationship between residing in any enhancer and effect size?



```{r}

pvals.sig <- pvals.lifted.gr[pvals.lifted.gr$lfdr.from.ss <= alpha]

# Break it up
breaks <- seq(from=0.01, to = 0.075, by = 0.005)
bins.ix <- .bincode(abs(pvals.sig$pi.diff), breaks = breaks)
bins.ix[is.na(bins.ix)]  <- -1

# Use already built functions to (not) filter and convert
#
tmp <- promcap.3 %>% dplyr::filter(mean.chicago > 10)
promcap.filt.gr <- filter_by_cell_type(tmp, "EP", 0)

#
percent.by.effect.df <-  matrix(NA, nrow = length(breaks), ncol = 3)

for (ii in 1:(length(breaks)-1)){
  sub <- pvals.sig[(ii == bins.ix)]
  tmp <- tally_overlaps(sub, promcap.filt.gr)
  
  # Frame it as a Binomial experiment
  k.successes <- tmp[1]
  N.trials <- length(sub)
  
  tmp <- prop.test(x=k.successes, n=N.trials, conf.level = 0.95, correct = F)
  percent.by.effect.df[ii, 2:3] <- as.vector(tmp$conf.int)
  
  percent.by.effect.df[ii, 1] <- k.successes / N.trials
}


# Baseline
baseline.p <- tally_overlaps(pvals.lifted.gr, promcap.filt.gr)[1] / length(pvals.lifted.gr)

```

```{r}
p.by.e.df <- as.data.frame(percent.by.effect.df)
colnames(p.by.e.df) <- c("p", "lower", "upper")
p.by.e.df$effect <- breaks

p.by.e.df %>% 
  ggplot(aes(x = effect, y = p)) +
  geom_point(size = 3, color = "blue") +
  geom_errorbar(aes(ymin = lower, ymax = upper)) +
  theme_minimal() +
  xlab("Absolute effect size of DMP") +
  ylab("Probability of DMP\nresiding in enhancer") +
  geom_hline(yintercept = baseline.p, color = "grey30", linetype = "dashed")
```
What the fuck does this mean? 


# Genes with DM Enhancers

A gene can be linked to multiple enhancers. And one enhancer could be linked to multiple genes.

```{r}
# Sort of hacky. Filter whole dataset, then use
# already built function to convert to granges
tmp <- promcap.3 %>% dplyr::filter(med.chicago > 5)
promcap.filt.gr <- filter_by_cell_type(tmp, "EP", 0)

print(paste0("Number of interactions considered: ",  length(promcap.filt.gr)))

dmps.gr <- pvals.lifted.gr[pvals.lifted.gr$lfdr.from.ss <= alpha]

# 
links <- find_and_curate_overlaps(dmps.gr, promcap.filt.gr)
print(paste0("Number of interaction-DMP pairs: ",  nrow(links)))

```


How many unique genes? How many that are "valid" (i.e. protein coding)?

```{r}
u.genes <- get_unique_genes_from_overlaps(links)
length(u.genes)
```

How many that are "valid" (i.e. protein coding)?

Discrepancy arises from there being some linked genes that include 1+ protein coding and 1+ other. 

```{r}
sum(u.genes %in% valid.symbols)
```

# Gene Ontology On These

```{r}
u.genes <- get_unique_genes_from_overlaps(links)
u.pc.genes <- u.genes[u.genes %in% valid.symbols]

u.ids <- ensembldb::select(EnsDb.Hsapiens.v86, 
                                      keys= u.genes, 
                                      keytype = "SYMBOL", 
                                      columns = c("SYMBOL","GENEID"))

write.table(u.ids[ ,2], "dm-enhancer-genes.txt", quote = F, row.names = F)

library(goplot)


go.out <- goplot::run_gene_ontology(u.ids$GENEID)
```


```{r}
go.df <- go.out %>% 
  goplot::go_output_to_df() 


ix <- (go.df$Description == "negative regulation of antigen processing and presentation of peptide or polysaccharide antigen via MHC class II")

go.df$Description[ix] <- "negative regulation of antigen processing"

go.plot <- go.df %>% 
  goplot::plot_go_barchart(20)


cowplot::save_plot("../../Figs/PCHi-C/gene-ontology-pchic.png",go.plot, base_width = 12.5, base_height = 7)
```

# 

```{r}
links.summary <- links %>% 
  group_by(oe.id) %>% 
  summarize(linked.gene,
            med.pi.diff = median(pi.diff),
            n.dmps = n(),
            dist) %>% 
  distinct()
```

# Integrate with Differential Expression

```{r}
combined.df <- inner_join(links.summary, expression.genes, by=c("linked.gene"="gene_name"))

nrow(combined.df)

sum(combined.df$linked.gene %in% valid.symbols)
length(unique(combined.df$linked.gene))
```


# Rank Comparisons

# Integration With Other Data Sets

```{r}

library(ComplexHeatmap)
dm.enhancer.genes <- get_unique_genes_from_overlaps(links.summary)
de.genes <- expression.genes$gene_name
dm.genes <- read_csv("../../DataExport/2023-03-01-DMGenes-forReid.csv") %>% 
  dplyr::filter(lfdr < 0.01) %>% 
  dplyr::pull(gene.name)

```
```{r}
set.data <- list(GeneBody = dm.genes,
                 Enhancer = dm.enhancer.genes,
                 Expression = expression.genes$gene_name)

set.data.mat <- list_to_matrix(set.data)
set.counts <- make_comb_mat(set.data.mat)
set.counts
```
```{r}
UpSet(set.counts)
```
What do we make of this? What information do we gain?

- Comparatively many shared genes in enhancer vs dm gene body. Really need to look at just promoters.
